src/protocols/echo/CHIPEchoClient.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements an object for a CHIP Echo unsolicitied
+ *      initiator (client).
+ *
+ */
+
+#include "CHIPEcho.h"
+
+namespace chip {
+namespace Protocols {
+
+ChipEchoClient::ChipEchoClient()
+{
+    ExchangeMgr            = nullptr;
+    OnEchoResponseReceived = nullptr;
+    ExchangeCtx            = nullptr;
+}
+
+CHIP_ERROR ChipEchoClient::Init(ExchangeManager * exchangeMgr)
+{
+    // Error if already initialized.
+    if (ExchangeMgr != nullptr)
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    ExchangeMgr            = exchangeMgr;
+    OnEchoResponseReceived = nullptr;
+    ExchangeCtx            = nullptr;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ChipEchoClient::Shutdown()
+{
+    if (ExchangeCtx != nullptr)
+    {
+        ExchangeCtx->Abort();
+        ExchangeCtx = nullptr;
+    }
+
+    ExchangeMgr = nullptr;
+
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * Send an echo request to a CHIP node.
+ *
+ * @param nodeId        The destination's nodeId
+ * @param payload       A System::PacketBuffer with the payload. This function takes ownership of the System::PacketBuffer
+ *
+ * @return CHIP_ERROR_NO_MEMORY if no ExchangeContext is available.
+ *         Other CHIP_ERROR codes as returned by the lower layers.
+ */
+CHIP_ERROR ChipEchoClient::SendEchoRequest(uint64_t nodeId, System::PacketBuffer * payload)
+{
+    // Discard any existing exchange context. Effectively we can only have one Echo exchange with
+    // a single node at any one time.
+    if (ExchangeCtx != nullptr)
```

Comment:
what is the purpose of the `Init()` method if we discard the context on every echo request?

Reply:
Init() is used to set ExchangeManager, not ExchangeContext.